JavaWeb3/src/main/resources/templates/sign_up.html

```diff
@@ -37,6 +37,49 @@
 	    new Vue({
             el: '#app',
             data: function() {
+				var validateEmail = (rule, value, callback) => {
+					var reg = /^[A-Za-z0-9_\-\.]+\@[A-Za-z0-9_\-\.]+\.([A-Za-z]{2,6})$/; 
+								//E-mail contains only one '@' character
+								//E-mail contains at least a dot ('.')
+								//E-mail contains between 2 and 6 characters after a dot ('.')
+					if (value === '') {
+						callback();
+					} else if (!reg.exec(value)) {
+						callback(new Error('Please input a valid email address.'));
+					}
+				};
+				var validateUsername = (rule, value, callback) => {
+					var reg1 = /[^a-zA-Z0-9_\-\.]/; //Username can only contain number, character and ., -, _
```

Comment:
A `^` symbol inside a set means the negation of that set, so it will accept everything but what you have in the set. For the validation of usernames I have used the following regex: `\b[a-zA-Z][a-zA-Z0-9\-._]{3,30}\b`

Reply:
Yes, I know, so this regexp together with conditional statement means if any character not in this set appears, emit an error message. And I dont want to show a lengthy notice (and also not sure how to construct), so I break down the expression into several.
And I only implemented validation. Do you think we should filter input and continue forwarding, or just refuse like what I've done?

Reply:
Breaking down into multiple expressions to verify is a great way to determine where the problem is coming from.
Now I get what you mean about the conditional statement, and it is right. It passed the unit tests for `username` validation.

Reply:
I can write a unit test and make another pull request. What do you think?

Reply:
Sure, you can use the unit tests from [here.](https://github.com/MetallicFocus/team_zero_application/blob/master/Android/app/src/test/java/teamzero/chat/mobile/RegistrationUnitTests.java)